File: nqueen.cpp
#include <bits/stdc++.h>
using namespace std;

bool isSafe(int** arr, int x, int y, int n) {
    // Check the column
    for (int row = 0; row < x; row++) {
        if (arr[row][y] == 1) {
            return false;
        }
    }

    // Check the upper left diagonal
    int row = x, col = y;
    while (row >= 0 && col >= 0) {
        if (arr[row][col] == 1) {
            return false;
        }
        row--;
        col--;
    }

    // Check the upper right diagonal
    row = x, col = y;
    while (row >= 0 && col < n) {
        if (arr[row][col] == 1) {
            return false;
        }
        row--;
        col++;
    }

    return true;
}

bool solvequeen(int** arr, int x, int n) {
    if (x >= n) {
        return true;  // All queens are placed
    }

    for (int col = 0; col < n; col++) {
        if (isSafe(arr, x, col, n)) {
            arr[x][col] = 1;  // Place queen

            if (solvequeen(arr, x + 1, n)) {
                return true;
            }

            arr[x][col] = 0;  // Backtrack
        }
    }
    return false;
}

void print(int** arr, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cout << (arr[i][j] == 1 ? "Q" : ".") << " ";
        }
        cout << endl;
    }
}

int main() {
    int n;
    cout << "Enter number of n: ";
    cin >> n;

    int** arr = new int*[n];
    for (int i = 0; i < n; i++) {
        arr[i] = new int[n]();
    }

    if (solvequeen(arr, 0, n)) {
        print(arr, n);
    } else {
        cout << "Solution does not exist" << endl;
    }

    // Clean up memory
    for (int i = 0; i < n; i++) {
        delete[] arr[i];
    }
    delete[] arr;

    return 0;
}

File: nqueen.py
def issafe(grid, x, y, n):
    # Check the column for any queen
    for row in range(n):
        if grid[row][y] == 1:
            return False

    # Check the upper left diagonal
    row = x
    col = y
    while row >= 0 and col >= 0:
        if grid[row][col] == 1:
            return False
        row -= 1
        col -= 1

    # Check the upper right diagonal
    row = x
    col = y
    while row >= 0 and col < n:
        if grid[row][col] == 1:
            return False
        row -= 1
        col += 1

    return True


def nqueen(grid, x, n):
    # Base case: when x equals n, a valid solution has been found
    if x == n:
        # Print the solution in a grid format
        for row in grid:
            print(row)
        print()  # Print a newline after each solution
        return

    # Try placing a queen in each column of row x
    for y in range(n):
        if issafe(grid, x, y, n):
            grid[x][y] = 1  # Place the queen
            nqueen(grid, x + 1, n)  # Recur to place rest of the queens
            grid[x][y] = 0  # Backtrack


n = int(input("Enter the number of queens: "))
if n in [4, 8, 16, 32]:
    grid = [[0 for j in range(n)] for i in range(n)]
    nqueen(grid, 0, n)
else:
    print("No solution")
